crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -122,4 +135,239 @@ module sui::random {
+        while (rr.max_size < rr.latest_round - rr.oldest_round + 1) {
+            dynamic_field::remove<u64, vector<u8>>(&mut rr.id, rr.oldest_round);
+            rr.oldest_round = rr.oldest_round + 1;
+        };
+        dynamic_field::add(&mut rr.id, round, value);
+    }
+
+    /// Set of inputs that can be used to create a RandomGenerator
+    // TODO: better name?
+    struct RandomnessRequest has store {
+        round: u64,
+        seed: vector<u8>,
+    }
+
+    public fun create_randomness_request(rr: &RandomRounds, ctx: &mut TxContext): RandomnessRequest {
+        RandomnessRequest {
+            round: rr.latest_round + 2, // next round that is safe when the current transaction is executed.
+            seed: to_bytes(fresh_object_address(ctx)), // globally unique (thuogh predictable).
+        }
+    }
+
+    /// Deterministic derivation of the seed for the given round.
+    public fun fulfill_and_create_generator(req: &RandomnessRequest, rr: &RandomRounds): RandomGenerator {
+        assert!(req.round >= rr.oldest_round && req.round <= rr.latest_round, 0); // TODO: better error
+        let randomness = dynamic_field::borrow(&rr.id, req.round);
+        let seed = hmac_sha3_256(randomness, &req.seed);
+        RandomGenerator {
+            seed,
+            counter: 0,
+            buffer: vector::empty(),
+        }
+    }
+
+    public fun is_available(req: &RandomnessRequest, rr: &RandomRounds): bool {
+        req.round >= rr.oldest_round && req.round <= rr.latest_round
+    }
+
+    public fun is_too_old(req: &RandomnessRequest, rr: &RandomRounds): bool {
+        req.round < rr.oldest_round
+    }
+
+    public fun required_round(req: &RandomnessRequest): u64 {
+        req.round
+    }
+
+    // TODO: Decide on how dapps could know when to call fulfill_and_create_generator.
+    // Options:
+    //  1. [Polling, Node APIs] tx1 calls create_randomness_request which returns the required randomness round either
+    //     as an output or as an event (or read directly from the chain)
+    //     -> dapp reads the value of RandomRounds from FN until the relevant randomness is available
+    //     -> sends tx2 that calls fulfill_and_create_generator
+    //  2. [Events] tx1 calls create_randomness_request which returns the required randomness round (as before)
+    //     -> waits for emitted event "round X is available" (by the random module)
+    //     -> sends tx2 that calls fulfill_and_create_generator
+    //  3. [Deferred execution for tx2] tx1 calls create_randomness_request
+    //     -> waits for tx1 finalization
+    //     -> sends tx2 that calls fulfill_and_create_generator with the next version of Random; this guarantees that
+    //     once tx2 is executed, the required randomness round is available.
+    // I think that 2 is the simplest to start with; if there is a need, one could implement a public "pusher service"
+    // for tx2.
+
+    ////////////////////////////////////
```

Comment:
RandomGenerator is implemented the same as in https://github.com/MystenLabs/sui/pull/15019